handlejoin.go

```diff
@@ -0,0 +1,278 @@
+package gomatrixserverlib
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/matrix-org/gomatrixserverlib/spec"
+	"github.com/matrix-org/util"
+)
+
+type HandleMakeJoinInput struct {
+	Context            context.Context
+	UserID             spec.UserID
+	RoomID             spec.RoomID
+	RoomVersion        RoomVersion
+	RemoteVersions     []RoomVersion
+	RequestOrigin      spec.ServerName
```

Comment:
Do we need the request origin or can we just extract this from the `UserID`? It would be a cleaner API as it would stop "known broken" configurations (i.e it is never valid to have `@foo:bar` and a `RequestOrigin` that isn't `bar`?

Reply:
Because this is the receiving side